sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -699,49 +504,73 @@ public Write withQueueDeclare(boolean queueDeclare) {
     private static class WriteFn extends DoFn<RabbitMqMessage, Void> {
       private final Write spec;
 
-      private transient ConnectionHandler connectionHandler;
+      private UUID writerId;
+      private transient ChannelLeaser channelLeaser;
 
       WriteFn(Write spec) {
         this.spec = spec;
       }
 
       @Setup
       public void setup() throws Exception {
-        connectionHandler = new ConnectionHandler(spec.uri());
-        connectionHandler.start();
-
-        Channel channel = connectionHandler.getChannel();
-
-        if (spec.exchange() != null && spec.exchangeDeclare()) {
-          channel.exchangeDeclare(spec.exchange(), spec.exchangeType());
+        if (writerId == null) {
+          writerId = UUID.randomUUID();
         }
-        if (spec.queue() != null && spec.queueDeclare()) {
-          channel.queueDeclare(spec.queue(), true, false, false, null);
+        if (channelLeaser == null) {
+          channelLeaser = spec.connectionHandlerProviderFn().apply(null);
         }
+
+        ChannelLeaser.UseChannelFunction<Void> setupFn =
+            (channel) -> {
+              if (spec.exchange() != null && spec.exchangeDeclare()) {
+                channel.exchangeDeclare(spec.exchange(), spec.exchangeType());
+              }
+              if (spec.queue() != null && spec.queueDeclare()) {
+                channel.queueDeclare(spec.queue(), true, false, false, null);
+              }
+              return null;
+            };
+
+        channelLeaser.useChannel(writerId, setupFn);
       }
 
       @ProcessElement
       public void processElement(ProcessContext c) throws IOException {
         RabbitMqMessage message = c.element();
-        Channel channel = connectionHandler.getChannel();
 
+        // TODO: get rid of 'exchange' vs 'queue'
         if (spec.exchange() != null) {
-          channel.basicPublish(
-              spec.exchange(),
-              message.getRoutingKey(),
-              message.createProperties(),
-              message.getBody());
+          ChannelLeaser.UseChannelFunction<Void> basicPublishFn =
+              (channel) -> {
+                channel.basicPublish(
+                    spec.exchange(),
+                    message.getRoutingKey(),
+                    message.createProperties(),
+                    message.getBody());
+                return null;
+              };
+
+          channelLeaser.useChannel(writerId, basicPublishFn);
         }
+
         if (spec.queue() != null) {
-          channel.basicPublish(
-              "", spec.queue(), MessageProperties.PERSISTENT_TEXT_PLAIN, message.getBody());
+
+          ChannelLeaser.UseChannelFunction<Void> basicPublishFn =
+              (channel) -> {
+                // TODO: what's with this ridiculous hard-coding of message props?
```

Comment:
![kill-it-with-fire](https://user-images.githubusercontent.com/145036/71425362-d37fa180-2669-11ea-90e8-ef5a5d2130d5.gif)